### PR TITLE
[Feature] Schedules

### PR DESCRIPTION
**Pull request by Civolilah**
Here are screenshots of implemented solution for schedules:

Screenshot 2023-02-24 at 01 57 25

Screenshot 2023-02-24 at 01 57 43

@turbo124 We have some keywords for this feature to add to the translation files: `schedules`, `email_statement`, `next_run`, `new_schedule`, `show_aging_table`, `show_payments_table`, `all_clients`, `edit_schedule`. I think those keywords are all we need to add.

@beganovich @turbo124 Let me know your thoughts.

### CODE REVIEW

File: src/components/forms/DebouncedCombobox.tsx [hunk 1]
```diff
@@ -161,6 +161,8 @@ export function DebouncedCombobox(props: Props) {
 
       if (props.clearInputAfterSelection) {
         setSelectedOption({ record: internalRecord, withoutEvents: true });
+
+        debouncedSearch('');
```
Comment (Civolilah) [inline src/components/forms/DebouncedCombobox.tsx:165]: The reason for adding this line of code is the bug we had when selecting values ​​when `clearInputAfterSelection` is true. When we selected one of them, it filtered list using value from selected item.

**Comment by turbo124** [comment]
@Civolilah looks good, a couple of small issues.

image

We also need to allow users to schedule the statements from the statement page

image

Please set endless as the default

When selecting a client from the dropdown list, the app crashes

**Comment by Civolilah** [comment]
> 

@turbo124 All of mentioned things are finished, now the user is able to schedule statement from the statement page, here is the screenshot:

Screenshot 2023-02-27 at 20 17 39

Also, `endless` is set by default for `remaining_cycles` property.

The only thing I couldn't find was an error in the client selection. I have indeed tested many versions to recreate the behavior but have not succeeded. Can you please just test it one more time and if happened again, just give me more details for this one?

File: src/common/hooks/permissions/useHasPermission.ts [hunk 2]
```diff
@@ -36,6 +36,10 @@ type PurchaseOrderPermissions =
   | 'create_purchase_order'
   | 'view_purchase_order'
   | 'edit_purchase_order';
+type TaskSchedulePermissions =
+  | 'create_task_schedule'
+  | 'view_task_schedule'
+  | 'edit_task_schedule';
```
Comment (Civolilah) [inline src/common/hooks/permissions/useHasPermission.ts:43]: @turbo124 Please let me know if those specified permissions are correct for schedules.

Comment (turbo124) [reply src/common/hooks/permissions/useHasPermission.ts:43]: @Civolilah this is admin/owner only.

Comment (Civolilah) [reply src/common/hooks/permissions/useHasPermission.ts]: @turbo124 Thankss, in that case I just changed conditions for enabling schedules queries.

**Comment by turbo124** [comment]
looks great!

**Comment by turbo124** [comment]
looks great!

File: src/pages/clients/statement/Statement.tsx [hunk 3]
```diff
@@ -62,15 +66,20 @@ export function Statement() {
 
   const dates = [
     {
-      id: 'last_7_days',
+      id: 'last7_days',
```
Comment (Civolilah) [inline src/pages/clients/statement/Statement.tsx:69]: This part has been changed to match date keys on schedule page for API request. This change does not have impact on functionality on Statement page.

File: src/components/forms/DebouncedCombobox.tsx (hunk 1 above)
Comment (beganovich) [reply src/components/forms/DebouncedCombobox.tsx:165]: Can you demo this using video or something? I am not sure we want to "clear after input selection"?

File: src/pages/settings/schedules/common/atoms.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * @copyright Copyright (c) 2022. Invoice Ninja LLC (https://invoiceninja.com)
+ *
+ * @license https://www.elastic.co/licensing/elastic-license
+ */
+
+import { Parameters } from 'common/interfaces/schedule';
+import { atom } from 'jotai';
+
+export const scheduleParametersAtom = atom<Parameters | undefined>(undefined);
```
Comment (beganovich) [inline src/pages/settings/schedules/common/atoms.ts:14]: Generally, there's no need for new files for such things. It's okay to export them in the same file as the component where the need for such a thing is first.

File: src/pages/settings/schedules/create/Create.tsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+          show_aging_table: false,
+          show_payments_table: false,
+          status: 'all',
+        },
+      });
+    }
+
+    return () => {
+      isProduction() && setParametersAtom(undefined);
+    };
```
Comment (beganovich) [inline src/pages/settings/schedules/create/Create.tsx:81]: We do not use this pattern anymore. Please check other resources /create page to see how initial data is set.

**Review by beganovich — CHANGES_REQUESTED** [review]
Looks good, small changes needed 👍

File: src/pages/settings/schedules/common/atoms.ts (hunk 4 above)
Comment (Civolilah) [reply src/pages/settings/schedules/common/atoms.ts:14]: Okay, thank you for advise. Now, this atom is exported from EmailStatement component where it is used.

File: src/pages/settings/schedules/create/Create.tsx (hunk 5 above)
Comment (Civolilah) [reply src/pages/settings/schedules/create/Create.tsx:81]: Okay, thanks! Than, this code has been slightly changed.

**Pull request merged**